File: test/language/templates/java/templates/InstantiateTypeAsChoiceFieldTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.instantiate_type_as_choice_field.InstantiateTypeAsChoiceField;
import templates.instantiate_type_as_choice_field.Test32;

public class InstantiateTypeAsChoiceFieldTest
{
    @Test
    public void readWrite() throws IOException
    {
        final InstantiateTypeAsChoiceField instantiateTypeAsChoiceField = new InstantiateTypeAsChoiceField(true);
        instantiateTypeAsChoiceField.setTest(new Test32(13));

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        instantiateTypeAsChoiceField.write(writer);

        final BitStreamReader reader = new ByteArrayBitStreamReader(
                writer.toByteArray(), writer.getBitPosition());
        InstantiateTypeAsChoiceField readInstantiateTypeAsChoiceField =
                new InstantiateTypeAsChoiceField(reader, true);
        assertTrue(instantiateTypeAsChoiceField.equals(readInstantiateTypeAsChoiceField));
    }
}

File: test/language/templates/cpp98/InstantiateTypeAsChoiceFieldTest.cpp
#include "gtest/gtest.h"

#include "templates/instantiate_type_as_choice_field/InstantiateTypeAsChoiceField.h"

namespace templates
{
namespace instantiate_type_as_choice_field
{

TEST(InstantiateTypeAsChoiceFieldTest, readWrite)
{
    InstantiateTypeAsChoiceField instantiateTypeAsChoiceField;
    instantiateTypeAsChoiceField.initialize(true);
    Test32 test32;
    test32.setValue(13);
    instantiateTypeAsChoiceField.setTest(test32);

    zserio::BitStreamWriter writer;
    instantiateTypeAsChoiceField.write(writer);
    size_t bufferSize = 0;
    const uint8_t* buffer = writer.getWriteBuffer(bufferSize);

    zserio::BitStreamReader reader(buffer, bufferSize);
    InstantiateTypeAsChoiceField readInstantiateTypeAsChoiceField(reader, true);

    ASSERT_TRUE(instantiateTypeAsChoiceField == readInstantiateTypeAsChoiceField);
}

} // namespace instantiate_type_as_choice_field
} // namespace templates
